libass/ass.c

```diff
@@ -683,6 +685,130 @@ static void event_format_fallback(ASS_Track *track)
+     *    Signature comment
+     *  56bc0a6736cdc7edab837ff8f304661fd16de0e4  2015.02.08
+     *    Allow custom PlayRes(X|Y)
+     *  a8ba2a2c1294a330a0e79ae7f0d3a203a7599166  2020.04.17
+     *    Set 'ScaledBorderAndShadow: yes'
+     *
+     * libav outputs initial ffmpeg format. (no longer maintained)
+     */
+
+    signed char conv_score = 0;
+    uint32_t tmp;
+
+    // If any headers not present in legacy ffmpeg were set, not legacy ffmpeg
+    tmp = track->parser_priv->sinfo_flags;
+    tmp <<= sizeof(tmp)*CHAR_BIT - SINFO_GENBY;
+    tmp >>= sizeof(tmp)*CHAR_BIT - SINFO_GENBY;
+    // Note: If there's SINFO_SCRIPTTYPE in the future this needs to be updated
+    tmp &= ~(1 << SINFO_PLAYRESX | 1 << SINFO_PLAYRESY);
+    if (tmp != 0)
+        goto not_legacy_ffmpeg;
+
+    // Legacy ffmpeg only ever has one style
+    // 2 not 1 because libass also adds a def style
+    if (track->n_styles != 2)
+        conv_score -= 3;
+    else if (!strncmp(track->styles[1].Name, "Default", 7))
+        conv_score += 3;
+
+    if (!strncmp(track->style_format,
+      "Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour,"
+      " BackColour, Bold, Italic, Underline, BorderStyle, Outline, Shadow,"
+      " Alignment, MarginL, MarginR, MarginV, AlphaLevel, Encoding",
+       379))
+        conv_score += 7;
+
+    if (!strncmp(track->event_format, "Layer, Start, End, Style, Text", 30)
+       || !strncmp(track->event_format, "Layer, Start, End, Text", 23))
+        conv_score += 4;
+
+    // Even though the GENBY signature may be set, we may, maybe have
+    // an ASS file based on an converted srt (dialouge only)
+    // where manual typesets requiring SBAS=0 were added afterwards
+    // If more than one style, and no custom format lines,
+    // assume this is the case
+    // If there are custom format lines, then VSFilter can't display
+    // the subs and even if there are some additions, to an legacy ffmpeg
+    // generated file would have been typeset with libass and SBAS=1 by default.
+    // so continue checking if legacy ffmpeg.
+    if (conv_score < 0)
+        goto not_legacy_ffmpeg;
+
+    if (sinfo_set(track->parser_priv, SINFO_GENBY)) {
+        if (!sinfo_set(track->parser_priv, GENBY_FFMPEG))
+            return;
+        else
+            goto set_legacy_ffmpeg;
+    }
+
+    // No definite clues: Continue to make an educated guess
+    // if this is an unmodified legacy ffmpeg file
+    if (track->track_type != TRACK_TYPE_ASS)
+        conv_score -= 5;
+
+    if ((track->PlayResX != 384 || track->PlayResY != 288)
+        && (
+            sinfo_set(track->parser_priv, SINFO_PLAYRESX)
+            || sinfo_set(track->parser_priv, SINFO_PLAYRESY)
+           )
+       )
+        conv_score -= 4;
+
+    //Final Score
+    ass_msg(track->library, MSGL_DBG2,
+            "Track legacy conv score: %d",
+            conv_score);
+
+    if (conv_score < 12)
+        goto not_legacy_ffmpeg;
+
+set_legacy_ffmpeg:
```

Comment:
I think that `detect` function should not set anything but return a boolean flag or the `conv_score` directly. All updates should be done outside of it. Also it'd be much cleaner without a ton of gotos.

Reply:
Done.